File: LeetCode/Java Solutions/Path Crossing/path.java
import java.util.HashSet;
import java.util.Set;

class Solution {
    public boolean isPathCrossing(String path) {
        // Initialize the starting point and the set of visited points.
        int x = 0, y = 0;
        Set<String> visited = new HashSet<>();
        visited.add(x + "," + y);

        for (char move : path.toCharArray()) {
            // Update the coordinates based on the direction.
            if (move == 'N') y++;
            else if (move == 'S') y--;
            else if (move == 'E') x++;
            else if (move == 'W') x--;

            // Create a string representation of the current coordinate.
            String coord = x + "," + y;

            // Check if the new position has already been visited.
            if (visited.contains(coord)) {
                return true;
            }

            // Mark the new position as visited.
            visited.add(coord);
        }

        // The path does not cross itself.
        return false;
    }
}

File: LeetCode/Python Solutions/Path Crossing/path.py
class Solution:
    def isPathCrossing(self, path: str) -> bool:
        # Initialize the starting point and the set of visited points.
        x, y = 0, 0
        visited = {(x, y)}

        # Map the directions to their corresponding coordinate changes.
        directions = {'N': (0, 1), 'S': (0, -1), 'E': (1, 0), 'W': (-1, 0)}

        for move in path:
            # Update the coordinates based on the direction.
            dx, dy = directions[move]
            x, y = x + dx, y + dy

            # Check if the new position has already been visited.
            if (x, y) in visited:
                return True

            # Mark the new position as visited.
            visited.add((x, y))

        # The path does not cross itself.
        return False
